### DISCUSSION TITLE
Need workaround for broken label gravity

### DISCUSSION THREAD

**Discussion by glendeni** [ImageMagick/ImageMagick#3425] [Help] [answered]
I am trying to produce a label with centered text and padding to the left/right.  To do that I have been adding spaces to both left and right and that has worked on my home computer running ImageMagick 6.9.1-2, using this command

convert -gravity center -pointsize 20 -fill black label:'    test   '  test.png`

but on a EC2 instance running 6.9.10-68, for the same command the text is _not_ centered, having white space on the right side but none on the left. So I'm looking for some workaround to achieve a centered text with padding in a different manner.

Jack

**Comment by snibgo** [comment]
Adding spaces at the start and end used to work, but sadly no longer. Perhaps this is a bug that will be fixed.

```
magick -pointsize 20 label:"Hello world" -trim +repage -bordercolor White -border 5x2 x.png
```
This trims to remove all white space around the text, then adds a white border of 5 pixels to left and right, and 2 pixels to top and bottom.

**Reply by glendeni** [reply]
Thanks for your response, I will use it.  This has to be a bug since (1) a space is a valid character, just like any other - it is not 'null' (2) the _is_ white space to the right of the string.

**Comment by GeeMack** [comment, marked as answer]
Under some circumstances ImageMagick will drop leading spaces in labels. The workaround is to escape the first space with a backslash "\\" like this...

    convert -gravity center -pointsize 20 -fill black label:'\ test ' test.png

The issue occurs on my versions 6.8.9 in a bash shell, and up to 6.9.11 and 7.0.11 on Windows. Adding the backslash appears to fix it in every case.

**Comment by glendeni** [comment]
Am back with a related problem.  The label text is automatically determined in my script, but its fallback is simply to have a blank label.  For the older version that was not a problem, but now if I use a string of only one (or more) spaces convert fails with message: 

_convert: no images defined `test.png' @ error/convert.c/ConvertImageCommand/3235_

So using border will not work in such a case and using a backslash appears to the universal solution.
(Or doing something kludgey like setting a text character but then setting the fill to the background color.)

**Reply by glendeni** [reply]
PS: FWIW using a backslash in text does cause other problems in a script which then need to be worked around, e.g. in my case using that text in names passed to a command-line or to name a file

**Comment by snibgo** [comment]
Thanks, @GeeMack. I hadn't thought of trying that. It works fine on Windows. If we want multiple leading spaces, we need to escape only the first one. There is no need to escape trailing spaces.
```
magick -pointsize 20 label:"\    Hello world" x.png
```
Yes, a space-only label would be:
```
magick -pointsize 20 label:"\ " x.png
```
